File: src/main/kotlin/uk/ac/ic/matterialize/matting/backgroundcut/models/ColorModel.kt
package uk.ac.ic.matterialize.matting.backgroundcut

import org.opencv.core.Core
import org.opencv.core.CvType
import org.opencv.core.Mat
import org.opencv.core.Scalar

class ColorModel(private val bgImage: Image) {
    private val globalBgModel = GlobalBgModel(bgImage.flattened)
    private val pixelBgModel = PixelBgModel(bgImage)
    private val fgModel = FgModel(pixelBgModel)
    private var prevSegResult: SegmentationResult = SegmentationResult.empty(bgImage.mat.rows(), bgImage.mat.cols())

    // alpha in equation (4)
    // must be between 0 and 1
    private val mixingFactor = 0.5

    init {
        assert(mixingFactor in 0.0..1.0)
    }

    fun updatePrevSegResult(prevSegmentationResult: SegmentationResult) {
        prevSegResult = prevSegmentationResult
    }

    fun mixProbs(image: FlatImage): Probs {
        val result = Mat()

        // equation (4)
        Core.addWeighted(
            globalBgModel.globalProbs(image).mat, mixingFactor,
            pixelBgModel.perPixelProbs(image).mat, 1 - mixingFactor, 0.0, result, CvType.CV_64FC1
        )
        return Probs(result)
    }

    fun colorTerms(image: FlatImage): ColorTerm {
        val bgEnergiesMat = Mat()
        Core.log(mixProbs(image).mat, bgEnergiesMat)
        Core.multiply(bgEnergiesMat, Scalar(-1.0), bgEnergiesMat)

        val fgEnergiesMat = Mat()
        Core.log(fgModel.globalProbs(image, prevSegResult).mat, fgEnergiesMat)
        Core.multiply(fgEnergiesMat, Scalar(-1.0), fgEnergiesMat)

        val bgEnergies = DoubleArray(image.numPixels())
        val fgEnergies = DoubleArray(image.numPixels())
        bgEnergiesMat.get(0, 0, bgEnergies)
        fgEnergiesMat.get(0, 0, fgEnergies)

        return ColorTerm(bgEnergies, fgEnergies)
    }
}

File: src/matting/background_cut/models/color_model.cpp
#include "color_model.hpp"
#include <opencv2/core.hpp>

using namespace cv;

Probability ColorModel::mix_probs(const Image &img) const {
  Mat result{};

  addWeighted(
      global_bg_model.global_probs(img).mat, mix_factor,
      pixel_bg_model.get_pixel_probability(img).mat, 1 - mix_factor, 0.0,
      result, CV_64FC1);

  return Probability(std::move(result));
}
ColorTerm ColorModel::color_term(const Image &img, const SegmentationResult &prev_segmentation_result) {
  Mat bgEnergies{};
  Mat fgEnergies{};

  log(mix_probs(img).mat, bgEnergies);
  multiply(bgEnergies, Scalar(-1.0), bgEnergies);

  log(global_fg_model.global_probs(img, prev_segmentation_result).mat, fgEnergies);
  multiply(fgEnergies, Scalar(-1.0), fgEnergies);

  return ColorTerm(std::move(bgEnergies), std::move(fgEnergies));
}
